### PR TITLE
ramips: add support for Nexx WT3020-16M MOD

### PR DESCRIPTION
**Pull request by Wingman2k**
This patch adds support to build a 16MB image for devices with modded flash and uboot.
Flash mod is easy to perform, but to build an Image patching and recompiling is needed.
On TL-MR3020 the size is detected automatically, don't know if this would be possible here too.
But with a fixed size an Image can be build larger than 8M and is ready to use.
This helps a lot to have a ready to use AIO travel Router and set up something.

Signed-off-by: Stefan Grau <stefan_grau@t-online.de>

### CODE REVIEW

File: target/linux/ramips/dts/WT3020-16M.dts [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/dts-v1/;
+
+#include "mt7620n.dtsi"
+
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "wt3020", "ralink,mt7620n-soc";
+	model = "Nexx WT3020 (8M)";
```
Comment (obsy) [inline target/linux/ramips/dts/WT3020-16M.dts]: This is typo? 16MB should be.

Also, please squash this in one commit

**Comment by pepe2k** [comment]
Hello @Wingman2k,

> This patch adds support to build a 16MB image for devices with **modded flash and uboot**.

We usually don't accept code which adds support for devices which need this kind of hardware modification and I don't think we can make exception here (yes, this means NAK from me).

> On TL-MR3020 the size is detected automatically, don't know if this would be possible here too.

Take a look here: https://bugs.lede-project.org/index.php?do=details&task_id=667#comment2292

Cheers,
Piotr

**Comment by bobafetthotmail** [comment]
> Take a look here: https://bugs.lede-project.org/index.php?do=details&task_id=667#comment2292

+1 for this. This would allow to merge the 4MiB and 8MiB images for Nexx WT3020 (as there are 2 versions currently that differ only in flash size) while also supporting flashmodded ones.
In these devices the wifi factory partition is before kernel/rootfs partitions anyway so you can expand the rootfs without issues.

**Comment by pepe2k** [comment]
@bobafetthotmail

> +1 for this. This would allow to merge the 4MiB and 8MiB images for Nexx WT3020 (as there are 2 versions currently that differ only in flash size) while also supporting flashmodded ones.
> In these devices the wifi factory partition is before kernel/rootfs partitions anyway so you can expand the rootfs without issues.

The only problem with this is that this approach (using MTDPART_SIZ_FULL binding) was rejected in upstream, just read lated comment from Mathias, for the same bug report.

Cheers,
Piotr

**Comment by Wingman2k** [comment]
Merging 4MiB and 8MiB would probably also restrict the image size to 4MB, to be compatible,
not really wanted this. I also totally understand to not add Modded devices (expected it).
there is "BLOCKSIZE := 4k" specified for 4MiB in source/target/linux/ramips/image/mt7620.mk

**Comment by mkresin** [comment]
>We usually don't accept code which adds support for devices which need this kind of hardware modification and I don't think we can make exception here (yes, this means NAK from me).

NAK from me as well for the very same reason.

**Comment by dony71** [comment]
I try this patch to support 16M flash chip, but it seem doesn't work
serial port dump log looks like below
Anybody help me what's wrong?

[quote]
Boot and Recovery Environment for Embedded Devices
Copyright (C) 2017 HackPascal <hackpascal@gmail.com>
Build date 2017-09-21 [git-35a696e]
Version 1.1 (r1126)

DRAM: 64MB
Platform: MediaTek MT7620N ver 2, eco 6
Board: Reference design
Clocks: CPU: 580MHz, Bus: 193MHz
Flash: Winbond W25Q128 (16MB) on rt2880-spi
rt2880-eth: Using MAC address 00:13:13:11:d2:bb
eth0: MediaTek MT7620N built-in 5-port 10/100M switch

Network started on eth0, inet addr 192.168.1.1, netmask 255.255.255.0

Press any key to interrupt autoboot ... 0

Trying to boot firmware from 0x00050000 in flash bank 0 ...
Reading data into memory ...
U-Boot firmware image header detected.
    Image Name:   MIPS LEDE Linux-4.9.58
    Data Size:    1356615 Bytes
    Load Address: 80000000
    Entry Point:  80000000
Uncompressing data (LZMA) ... done.
Flushing cache ... done.

Starting kernel at 0x80000000...

[    0.000000] Linux version 4.9.58 (user@ubuntuvm) (gcc version 5.5.0 (LEDE GCC 5.5.0 r5399-6aa4b97) ) #0 Tue Nov 21 14:15:20 2017
[    0.000000] Board has DDR1
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620N ver:2 eco:6
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is Nexx WT3020 (16M)
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 256 (order: -2, 1024 bytes)
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=00033a20
[    0.000000] Readback ErrCtl register=00033a20
[    0.000000] Memory: 60444K/65536K available (3155K kernel code, 153K rwdata, 756K rodata, 152K init, 216K bss, 5092K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS:256
[    0.000000] CPU Clock: 580MHz
[    0.000000] clocksource: systick: mask: 0xffff max_cycles: 0xffff, max_idle_ns: 583261500 ns
[    0.000000] systick: enable autosleep mode
[    0.000000] systick: running - mult: 214748, shift: 32
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000011] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.015453] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.080604] pid_max: default: 32768 minimum: 301
[    0.089915] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.102941] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.122131] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.141607] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.153752] pinctrl core: initialized pinctrl subsystem
[    0.164717] NET: Registered protocol family 16
[    0.194633] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.205649] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.218246] rt2880_gpio 10000660.gpio: registering 32 gpios
[    0.229232] rt2880_gpio 10000660.gpio: registering 32 irq handlers
[    0.241774] rt2880_gpio 10000688.gpio: registering 1 gpios
[    0.252581] rt2880_gpio 10000688.gpio: registering 1 irq handlers
[    0.268357] clocksource: Switched to clocksource systick
[    0.280081] NET: Registered protocol family 2
[    0.289550] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.303322] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.315915] TCP: Hash tables configured (established 1024 bind 1024)
[    0.328642] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.340172] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.352853] NET: Registered protocol family 1
[    0.363640] rt-timer 10000100.timer: maximum frequency is 1220Hz
[    0.376288] Crashlog allocated RAM at address 0x3f00000
[    0.387688] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    0.408699] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.420173] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.447858] io scheduler noop registered
[    0.455554] io scheduler deadline registered (default)
[    0.466017] ralink-usb-phy usbphy: invalid resource
[    0.476168] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.489814] console [ttyS0] disabled
[    0.496800] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 20, base_baud = 2500000) is a Palmchip BK-3103
[    0.516569] console [ttyS0] enabled
[    0.516569] console [ttyS0] enabled
[    0.530349] bootconsole [early0] disabled
[    0.530349] bootconsole [early0] disabled
[    0.550444] spi spi0.0: force spi mode3
[    0.558898] m25p80 spi0.0: w25q128 (16384 Kbytes)
[    0.568306] 4 ofpart partitions found on MTD device spi0.0
[    0.579277] Creating 4 MTD partitions on "spi0.0":
[    0.588834] 0x000000000000-0x000000030000 : "u-boot"
[    0.600365] 0x000000030000-0x000000040000 : "u-boot-env"
[    0.612811] 0x000000040000-0x000000050000 : "factory"
[    0.624742] 0x000000050000-0x000001000000 : "firmware"
[    2.001302] 2 uimage-fw partitions found on MTD device firmware
[    2.013137] 0x000000050000-0x00000019b387 : "kernel"
[    2.024660] 0x00000019b387-0x000001000000 : "rootfs"
[    2.036403] mtd: device 5 (rootfs) set to be root filesystem
[    2.048289] 1 squashfs-split partitions found on MTD device rootfs
[    2.060677] 0x000000ca4000-0x000001000000 : "rootfs_data"
[    2.073743] libphy: Fixed MDIO Bus: probed
[    2.085550] mtk_soc_eth 10100000.ethernet: loaded mt7620 driver
[    2.098201] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    2.115214] rt2880_wdt 10000120.watchdog: Initialized
[    2.126430] NET: Registered protocol family 10
[    2.138278] NET: Registered protocol family 17
[    2.147253] 8021q: 802.1Q VLAN Support v1.8
[    2.172513] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    2.187418] Freeing unused kernel memory: 152K
[    2.196296] This architecture does not have kernel memory protection.
[    2.648309] random: fast init done
[    3.839201] init: Console is alive
[    3.846195] init: - watchdog -
[    7.503825] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    7.870869] usbcore: registered new interface driver usbfs
[    7.881998] usbcore: registered new interface driver hub
[    7.892711] usbcore: registered new device driver usb
[    7.908962] exFAT: Version 1.2.9
[    7.947710] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    7.970992] SCSI subsystem initialized
[    7.983732] ehci-platform: EHCI generic platform driver
[    8.004499] phy phy-usbphy.0: remote usb device wakeup disabled
[    8.016315] phy phy-usbphy.0: UTMI 16bit 30MHz
[    8.025202] ehci-platform 101c0000.ehci: EHCI Host Controller
[    8.036691] ehci-platform 101c0000.ehci: new USB bus registered, assigned bus number 1
[    8.052595] ehci-platform 101c0000.ehci: irq 26, io mem 0x101c0000
[    8.073188] ehci-platform 101c0000.ehci: USB 2.0 started, EHCI 1.00
[    8.086771] hub 1-0:1.0: USB hub found
[    8.094621] hub 1-0:1.0: 1 port detected
[    8.105808] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    8.119750] ohci-platform: OHCI generic platform driver
[    8.130472] ohci-platform 101c1000.ohci: Generic Platform OHCI controller
[    8.144073] ohci-platform 101c1000.ohci: new USB bus registered, assigned bus number 2
[    8.159970] ohci-platform 101c1000.ohci: irq 26, io mem 0x101c1000
[    8.198046] hub 2-0:1.0: USB hub found
[    8.205974] hub 2-0:1.0: 1 port detected
[    8.220697] usbcore: registered new interface driver usb-storage
[    8.233636] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    8.252185] init: - preinit -
[   10.150596] 8021q: adding VLAN 0 to HW filter on device eth0
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[   11.435150] mount_root: loading kmods from internal overlay
[   11.501603] kmodloader: loading kernel modules from //etc/modules-boot.d/*
[   11.517272] kmodloader: done loading kernel modules from //etc/modules-boot.d/*
[   14.199197] jffs2: notice: (372) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   14.230790] block: attempting to load /tmp/jffs_cfg/upper/etc/config/fstab
[   14.252860] block: extroot: not configured
[   14.734395] jffs2: notice: (369) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   16.348044] block: attempting to load /tmp/jffs_cfg/upper/etc/config/fstab
[   16.368344] block: extroot: not configured
[   16.377910] mount_root: switching to jffs2 overlay
[   16.452162] urandom-seed: Seeding with /etc/urandom.seed
[   16.593348] procd: - early -
[   16.599184] procd: - watchdog -
[   17.299743] procd: - watchdog -
[   17.311672] procd: - ubus -
[   18.052937] procd: - init -
Please press Enter to activate this console.
[   18.521968] kmodloader: loading kernel modules from /etc/modules.d/*
[   18.682274] ntfs: driver 2.1.32 [Flags: R/O MODULE].
[   18.707840] ip6_tables: (C) 2000-2006 Netfilter Core Team
[   18.734233] fuse init (API version 7.26)
[   18.750458] usbcore: registered new interface driver cdc_wdm
[   18.764014] Loading modules backported from Linux version wt-2017-11-01-0-gfe248fc2c180
[   18.780029] Backport generated by backports.git v4.14-rc2-1-31-g86cf0e5d
[   18.797904] ip_tables: (C) 2000-2006 Netfilter Core Team
[   18.820259] nf_conntrack version 0.5.0 (1024 buckets, 4096 max)
[   18.835601] nf_conntrack_rtsp v0.7 loading
[   18.906770] nf_nat_rtsp v0.7 loading
[   18.919612] usbcore: registered new interface driver ums-alauda
[   18.933899] usbcore: registered new interface driver ums-cypress
[   18.948290] usbcore: registered new interface driver ums-datafab
[   18.962578] usbcore: registered new interface driver ums-freecom
[   18.976972] usbcore: registered new interface driver ums-isd200
[   18.991327] usbcore: registered new interface driver ums-jumpshot
[   19.005782] usbcore: registered new interface driver ums-karma
[   19.020343] usbcore: registered new interface driver ums-sddr09
[   19.034661] usbcore: registered new interface driver ums-sddr55
[   19.049164] usbcore: registered new interface driver ums-usbat
[   19.072549] usbcore: registered new interface driver usbserial
[   19.084354] usbcore: registered new interface driver usbserial_generic
[   19.097508] usbserial: USB Serial support registered for generic
[   19.139571] xt_time: kernel timezone is -0000
[   19.151113] usbcore: registered new interface driver cdc_ether
[   19.167025] usbcore: registered new interface driver cdc_ncm
[   19.206415] usbcore: registered new interface driver ch341
[   19.217529] usbserial: USB Serial support registered for ch341-uart
[   19.232350] usbcore: registered new interface driver cp210x
[   19.243650] usbserial: USB Serial support registered for cp210x
[   19.258872] usbcore: registered new interface driver ftdi_sio
[   19.270507] usbserial: USB Serial support registered for FTDI USB Serial Device
[   19.287586] usbcore: registered new interface driver huawei_cdc_ncm
[   19.354127] usbcore: registered new interface driver pl2303
[   19.365433] usbserial: USB Serial support registered for pl2303
[   19.382040] PPP generic driver version 2.4.2
[   19.393456] NET: Registered protocol family 24
[   19.405418] usbcore: registered new interface driver qmi_wwan
[   19.420342] usbcore: registered new interface driver rndis_host
[   19.448929] usbcore: registered new interface driver sierra
[   19.460226] usbserial: USB Serial support registered for Sierra USB modem
[   19.477280] usbcore: registered new interface driver sierra_net
[   19.492073] usbcore: registered new interface driver cdc_mbim
[   19.527890] rt2800_wmac 10180000.wmac: loaded eeprom from mtd device "factory"
[   19.542377] ieee80211 phy0: rt2x00_set_rt: Info - RT chipset 6352, rev 0500 detected
[   19.557828] ieee80211 phy0: rt2x00_set_rf: Info - RF chipset 7620 detected
[   19.611642] kmodloader: done loading kernel modules from /etc/modules.d/*
[   29.956123] 8021q: adding VLAN 0 to HW filter on device eth0
[   30.007338] br-lan: port 1(eth0.1) entered blocking state
[   30.018168] br-lan: port 1(eth0.1) entered disabled state
[   30.029398] device eth0.1 entered promiscuous mode
[   30.038995] device eth0 entered promiscuous mode
[   30.257749] br-lan: port 1(eth0.1) entered blocking state
[   30.268571] br-lan: port 1(eth0.1) entered forwarding state



BusyBox v1.27.2 () built-in shell (ash)

     _________
    /        /\      _    ___ ___  ___
   /  LE    /  \    | |  | __|   \| __|
  /    DE  /    \   | |__| _|| |) | _|
 /________/  LE  \  |____|___|___/|___|                      lede-project.org
 \        \   DE /
  \    LE  \    /  -----------------------------------------------------------
   \  DE    \  /    Reboot (SNAPSHOT, r5399-6aa4b97)
    \________\/    -----------------------------------------------------------

=== WARNING! =====================================
There is no root password defined on this device!
Use the "passwd" command to set up a new password
in order to prevent unauthorized SSH logins.
--------------------------------------------------
root@LEDE:/#
[/quote]

**Comment by bobafetthotmail** [comment]
that console dump looks fine, can you describe better the issue? what is not working?

**Comment by dony71** [comment]
client on LAN cannot get IP, even if i force, still cannot ping
i compare 8M dump vs 16M dump and notice rootfs_data is so big on 16M
is it normal?
0x000000ca4000-0x000001000000 : "rootfs_data"    >>> 16M flash
0x0000007b1000-0x000000800000 : "rootfs_data"   >>> 8M flash

**Comment by dony71** [comment]
i found the problem because patch creates new device wt3020-16M
however it's missing base configuration for that device
need to add wt3020-16M definition in 
target/linux/ramips/base-file/etc/diag.sh
target/linux/ramips/base-file/etc/board.d/01_leds
target/linux/ramips/base-file/etc/board.d/02_network

**Pull request closed without merging**